File: Python/max-points-on-a-line.py
from collections import defaultdict
from decimal import getcontext, Decimal
getcontext().prec = 25
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        
        def getLine(pt1, pt2):
            x1, y1 = pt1
            x2, y2 = pt2
                
            m = Decimal(y2-y1)/Decimal(x2-x1) if x1 != x2 else float('inf')
            c = y1 - m * x1 if x1 != x2 else x1
            
            return m, c
        
        # points.sort()
        n = len(points)

        if n < 3:
            return n
        
        curmax = 0
        
        # line represented as gradient, y intercept
        for i, pt1 in enumerate(points[:-1]):
            lines_i = defaultdict(int)
            duplicate = 0
            maxlines = 0
            for pt2 in points[(i+1):]:
                if pt2 == pt1:
                    duplicate += 1
                    continue
                line = getLine(pt1, pt2)
                lines_i[line] += 1
                maxlines = max(maxlines, lines_i[line])
            curmax = max(curmax, maxlines + duplicate + 1)
                
        return curmax

File: C++/max-points-on-a-line.cpp
class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        int n = points.size();
        if(n < 2) return n;
        vector<vector<int>> dp(n, vector<int>(n, 2));
        //dp[i][j] max number of points on the line of pint i and j
        
        int max = 2;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < n; j++)
            {
                if(j == i || dp[i][j] != 2) continue;
                vector<int> p1 = points[i];
                vector<int> p2 = points[j];
                vector<int> currSet = {i, j};
                for(int k = 0; k < n; k++)
                {
                    if(k == i || k == j) continue;
                    vector<int> p3 = points[k];
                    //if one the same line, push the point to the point set
                    if(((long)p3[0]-(long)p2[0])*((long)p2[1]-(long)p1[1]) 
                       == ((long)p2[0]-(long)p1[0])*((long)p3[1]-(long)p2[1]))
                        currSet.push_back(k);
                }
                //set all combinations to the size of the set
                if(currSet.size()!=2)
                {
                    for(int m = 0; m < currSet.size(); m++)
                    {
                         for(int n = 0; n < currSet.size(); n++)
                         {
                             if(m!=n) dp[currSet[m]][currSet[n]] = currSet.size();
                         }
                    }
                }
                max = std::max(max, (int)currSet.size());
            }
        }
        return max;
    }
};